### ISSUE TITLE
false negative with recursive type alias involving float sequences

### ISSUE THREAD

**Issue by jorenham** [KotlinIsland/basedmypy#819]
```py
from collections.abc import Sequence
from typing import TypeAlias

FloatVector: TypeAlias = Sequence[float]
FloatTensor: TypeAlias = FloatVector | Sequence["FloatTensor"]

rejected_float_vector: FloatVector = "ok"       # OK: rejected

accepted_float_tensor: FloatTensor = [[[3.14]]] # OK: accepted
rejected_float_vector1: FloatTensor = object()  # OK: rejected
rejected_float_vector2: FloatTensor = "fail"    # FAIL: accepted
```

I'm guessing this has to do with the `int` to `float` promotion stuff, as well as the fact that ...

> one sec, I have to scream into a pillow for a bit

...

> ... ok where was I 

... the fact that `str` is a `Sequence[str]`, which causes mypy to think that `float` is assignable to `str`. So I guess that makes total sense within the framework of "mypy logic".

**Comment by jorenham** [comment]
oh woops, this should be labeled as bug

**Comment by KotlinIsland** [comment]
> oh woops, this should be labeled as bug

feel free to update it 😁

**Comment by KotlinIsland** [comment]
is this the same thing?
```py
class C[T]: ...
type A = C[A] | int
type B = C[B]

b: B
a: A = b
```

**Comment by KotlinIsland** [comment]
-related #651

**Comment by jorenham** [comment]
> ```python
> class C[T]: ...
> type A = C[A] | int
> type B = C[B]
> 
> b: B
> a: A = b
> ```

If `T` is covariant, and mypy treats `str` like a recursive type, then it might indeed be a more general example.

**Comment by jorenham** [comment]
> is this the same thing?
> 
> ```python
> class C[T]: ...
> type A = C[A] | int
> type B = C[B]
> 
> b: B
> a: A = b
> ```

What's the `| int` for btw 🤔 ?

**Comment by KotlinIsland** [comment]
> What's the `| int` for btw 🤔 ?

just for being closer to the original egg

**Comment by jorenham** [comment]
this seems relevant

https://github.com/KotlinIsland/basedmypy/blob/0935784ab77e91c869bae67be01ba0f420678620/mypy/subtypes.py#L170-L189

**Comment by KotlinIsland** [comment]
i really don't understand this. if you could put together an egg, that would be good

**Comment by jorenham** [comment]
how about the egg in the issue itself?

**Comment by KotlinIsland** [comment]
> how about the egg in the issue itself?

i don't get that one 😅

probably an egg where it's not using any builtins, and there is a positive and a negative

**Comment by jorenham** [comment]
```py
class Increment[T]:  # covariant
    def current(self, /) -> T: ...  # type: ignore[empty-body]

type Infinity = Increment[Infinity]

type Zero = Increment[None]
type Integer = Zero | Increment[Integer]


def int2inf(integer: Integer, /) -> Infinity:
    return integer  # rejected => ok

def inf2int(infinity: Infinity, /) -> Integer:
    return infinity  # accepted => wut
```

- [mypy-play](https://mypy-play.net/?mypy=latest&python=3.12&gist=4211d58292b517604638e50b74e9fc69)
- [bpr-play](https://basedpyright.com/?pyrightVersion=1.22.0&typeCheckingMode=all&code=MYGwhgzhAECSB2wBOBTAti%2BAXA2gFQF0AuaaAYmmAHsA3MJASzGwFgAoU0gExQDNKArklTYAFBBQheAGmgB6AJTQAtAD5oeEgDod7dlgCeABxRx4vBvAaHoAXjPJ0mXAgtXDBPW0MnoALRQkKjsHVAxsHAA5KngUT29jUwQsFABzQJCAoOgAH1CnCOS0wPivHn5LLAAmS15RSuKkEiL0pFlFFXVXS2sDInZOaFQsIXhoBtbSClQAKxRgFK47dSoAazK%2BcfMasVqew2bzfYN2pTUzFNb%2Bjk5h0a23XqnoMGBgFCNF5egAdwEsdhAA)

**Comment by KotlinIsland** [comment]
does the covariance impact the situation?

is this the same thing?
```py
class Thing: ...
class Box[T]: ...

type CouldEnd = Box[CouldEnd] | Thing
type Recursive = Box[Recursive]

b: Recursive = Box()
a: CouldEnd = b # this should be an error

c: CouldEnd = Box[Box[Box[Thing]]]() # this is just for fun
```

**Comment by jorenham** [comment]
> does the covariance impact the situation?
> 
> is this the same thing?
> 
> ```python
> class Thing: ...
> class Box[T]: ...
> 
> type CouldEnd = Box[CouldEnd] | Thing
> type Recursive = Box[Recursive]
> 
> b: Recursive = Box()
> a: CouldEnd = b # this should be an error
> 
> c: CouldEnd = Box[Box[Box[Thing]]]() # this is just for fun
> ```

I thought that both mypy and pyright (incorrectly) inferred `T@Box` as covariant in such cases 🤔
